content/blog/posts/o3de-motion-matching-2022-04.md

```diff
@@ -0,0 +1,181 @@
+
+{{< image-width src="/images/blog/motion-matching/trajectory-history.png" width="100%" alt="Trajectory history samples showing the past positions and facing directions of the character." >}}
+
+## Controlling the character
+
+We talked about how the algorithm ensures that the animation is smooth but did not look into how we can actually guide and control our character yet - that is where the trajectory comes in. The motion matching algorithm not only takes a pose but also a desired future trajectory as input. It tries to synthesize an animation that matches the trajectory and makes the character walk onto, or close to, the desired path.
+
+{{< image-width src="/images/blog/motion-matching/script-canvas-graph.png" width="100%" alt="Script Canvas Graph for motion matching character control." >}}
+
+For an interactive game, we can use the gamepad's joystick position as a target to generate a desired future trajectory by creating an exponential curve starting at the current character position going towards the character facing direction, and then bend the curve towards the joystick's target position. This process is called trajectory prediction.
+
+{{< video src="/images/blog/motion-matching/trajectory-prediction.mp4" autoplay="true" loop="true" width="100%" info="Gamepad controlled character using motion matching." >}}
+<br>
+<br>
+
+## Data Analysis & Debugging
+
+While motion matching gets rid of the manual process of setting up locomotion in an animation graph, precisely timing transitions and making sure the blends look good, it also takes away some control. Visualizing data and the state of the algorithm and knowing our data is key for debugging an animation glitch.
+
+We provide several debug visualizations: Features like a joint velocity or the trajectory can be visualized together with the query values as well as the feature costs along with some performance metrics. The feature costs over time can be used to see if a visual glitch corresponds with a high cost at that given point in time. A high cost indicates that the algorithm struggled to find a good matching pose for the query. And this again says that our motion database might lack a motion for the given scenario that can be fixed by retaking or adding an extra animation clip.
```

Comment:
nit: theres not really a list here, so remove use of ':' and maybe focus just on key debug visualizations. "several debug visualizations" is a little passive. Maybe "O3DE provides a number of motion matching visualizations to aid troubleshooting. "You can display features, for example joint velocities or the trajectory...:"

Reply:
Changed